src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -476,7 +478,19 @@ private void ProcessEvents()
                         string expandedName = null;
                         WatchedDirectory associatedDirectoryEntry = null;
 
-                        if (nextEvent.wd != -1) // wd is -1 for events like IN_Q_OVERFLOW that aren't tied to a particular watch descriptor
+                        // Two events take priority: Overflow and SymLinks. An overflow event means that we can't trust 
+                        // our state without restarting since we missed events and some of those events could be a 
+                        // directory create, meaning we wouldn't have added the directory to the watch and would not provide
+                        // correct data to the caller. The other event, SymLinks, should be filtered out due to the FSW contract.
+                        if ((mask & (uint)Interop.libc.NotifyEvents.IN_Q_OVERFLOW) != 0)
+                        {
+                            // Notify the caller of the error and restart to pick up any potential directories we missed due
+                            // to the overflow.
+                            watcher.NotifyInternalBufferOverflowEvent();
+                            watcher.Restart();
```

Comment:
We don't restart like this on Windows, and it's not clear to me why we're doing it here.  Couldn't we miss additional events we wouldn't otherwise miss during the time between the stopping and restarting?  And even if we do want to do this restart, shouldn't it only be if IncludeSubdirectories, since the justification in the comment above implies we'd be doing this in case we missing a directory creation notification?